### ISSUE TITLE
Package export/import including necessary images

### ISSUE THREAD

**Issue by johnbelamaric** [GoogleContainerTools/kpt#2602]
### Describe your problem

Use case: As a packaged software vendor that may not make my images available publicly, I would like to be able to easily deliver a complete config and image bundle to my customers, and for them to be able to easily import it into their repository and registry.

### Proposed mechanism:

* A `pkg export` command that creates a single file that contains:
  - The `kpt` package
  - The images needed for any containers in that package
  - The images needed for any `kpt fn`s in that package
  
   The `pkg export` command would also need to rewrite the `image` fields of any resources or `kpt fn`s to indicate that the images referenced are stored locally in the archive.

* A `pkg import` command that imports a file created by `pkg export`. This would:
  - Push the `kpt` package to a git (or OCI?) repo
  - Push the container images to an image registry
  - Rewrite the `image` tags to reference that registry

See also https://github.com/GoogleContainerTools/kpt/pull/2589#discussion_r760346223

**Comment by johnbelamaric** [comment]
@bgrant0607 @mikebz @phanimarupaka

**Comment by phanimarupaka** [comment]
Thanks for the issue John. What is the priority of this issue ?

**Comment by johnbelamaric** [comment]
Backlog. No real urgency. 6-9 months?

**Comment by loudej** [comment]
Interesting. I for the kpt functions I believe there's also a `docker save` and `docker load` that can convert move an image between the local docker and tarball. So if you `docker build` the image to any name, you could then save and load on another machine. Image name wouldn't need to change --- you'd just need to be sure to render with never-pull-image option.

For the package itself... Maybe that wouldn't even need to be an OCI tar? If an export command did `docker save` to tar each function image to a nested folder - and then made a tar out of the package folder with images embedded - the reverse of untar and `docker load` on another system could be all you need?

**Comment by johnbelamaric** [comment]
Yes, good point with `docker save`. I forgot about it but have done that in the past. Looking at what I did then, I also retagged the images to remove the registry info. This made it easier to add a different registry later, and in that case sort of intentionally broke the chain back to the internal private registry. But it's not really needed. And as you said you don't even need to push the `kpt` function images to a new registry in this case.

This whole thing can just be a tarball, we just need to have conventions about how to identify what's in it so we can import it later. I don't know enough about the OCI structure to say if it's useful here or not.

**Comment by loudej** [comment]
Oh yes, true. Metadata would be a good reason to use OCI tar instead of a plain tar. For structure, if you're imagining a tar of layer tars + metadata jsons, I'd say that's a solid working assumption.

From how you described that, another thought, part of the operation could be useful outside of a bundle as well? A command to go over every kptfile in a package, pull-retag-push all images to private repo, and rewrite the kptfile image names? Not sure if people have ever asked for a utility to air-gap their rendering away from public function locations.
